crates/sui-types/src/crypto.rs

```diff
@@ -48,6 +48,116 @@ pub type AccountPublicKey = Ed25519PublicKey;
 pub type AccountPrivateKey = Ed25519PrivateKey;
 pub type AccountSignature = Ed25519Signature;
 
+#[enum_dispatch]
+#[derive(Debug)]
+pub enum SuiKeyPair {
+    Ed25519KeyPair,
+    Secp256k1KeyPair,
+}
+
+#[derive(Debug)]
+pub enum PublicKey<'a> {
+    Ed25519KeyPair(&'a Ed25519PublicKey),
+    Secp256k1KeyPair(&'a Secp256k1PublicKey),
+}
+
+impl SuiKeyPair {
+    pub fn public(&self) -> PublicKey<'_> {
+        match self {
+            SuiKeyPair::Ed25519KeyPair(kp) => PublicKey::Ed25519KeyPair(kp.public()),
+            SuiKeyPair::Secp256k1KeyPair(kp) => PublicKey::Secp256k1KeyPair(kp.public()),
+        }
+    }
+}
+
+impl signature::Signer<Signature> for SuiKeyPair {
+    fn try_sign(&self, msg: &[u8]) -> Result<Signature, signature::Error> {
+        match self {
+            SuiKeyPair::Ed25519KeyPair(kp) => kp.try_sign(msg),
+            SuiKeyPair::Secp256k1KeyPair(kp) => kp.try_sign(msg),
+        }
+    }
+}
+
+impl FromStr for SuiKeyPair {
+    type Err = eyre::Report;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let kp = Self::decode_base64(s).map_err(|e| eyre::eyre!("{}", e.to_string()))?;
+        Ok(kp)
+    }
+}
+
+impl EncodeDecodeBase64 for SuiKeyPair {
+    fn encode_base64(&self) -> String {
+        let mut bytes: Vec<u8> = Vec::new();
+        match self {
+            SuiKeyPair::Ed25519KeyPair(kp) => {
+                let kp1 = kp.copy();
+                bytes.extend_from_slice(&[self.public().flag()]);
+                bytes.extend_from_slice(kp.public().as_ref());
+                bytes.extend_from_slice(kp1.private().as_ref());
+            }
+            SuiKeyPair::Secp256k1KeyPair(kp) => {
+                let kp1 = kp.copy();
```

Comment:
this feels unclean, but needed a way to get both pubkey and privkey from keypair with privkey takes the ownership of the keypair object itself

happy to take suggestions

Reply:
I think it makes sense to copy